src/herder/ReadyEnvelopes.cpp

```diff
@@ -0,0 +1,141 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "herder/ReadyEnvelopes.h"
+#include "herder/Herder.h"
+#include "herder/HerderUtils.h"
+#include "main/Application.h"
+#include "scp/SCP.h"
+#include "util/Logging.h"
+
+#include <medida/counter.h>
+#include <medida/metrics_registry.h>
+
+namespace stellar
+{
+
+ReadyEnvelopes::ReadyEnvelopes(Application& app)
+    : mApp(app)
+    , mReadyEnvelopesSize(
+          mApp.getMetrics().NewCounter({"herder", "envelopes", "ready-size"}))
+    , mSeenEnvelopesSize(
+          mApp.getMetrics().NewCounter({"herder", "envelopes", "seen-size"}))
+{
+}
+
+ReadyEnvelopes::~ReadyEnvelopes()
+{
+}
+
+bool
+ReadyEnvelopes::seen(SCPEnvelope const& envelope)
+{
+    auto slotIndex = envelope.statement.slotIndex;
+    if (slotIndex < mMinimumSlotIndex)
+    {
+        return true;
+    }
+
+    auto& seen = mEnvelopes[slotIndex].mSeenEnvelopes;
+    return seen.find(envelope) != std::end(seen);
+}
+
+bool
+ReadyEnvelopes::push(SCPEnvelope const& envelope)
+{
+    if (seen(envelope))
+    {
+        return false;
+    }
+
+    traceEnvelope(mApp, "Marked ready", envelope);
+
+    auto slotIndex = envelope.statement.slotIndex;
+    assert(slotIndex >= mMinimumSlotIndex);
+    assert(mEnvelopes[slotIndex].mSeenEnvelopes.find(envelope) ==
+           std::end(mEnvelopes[slotIndex].mSeenEnvelopes));
+
+    mEnvelopes[slotIndex].mSeenEnvelopes.insert(envelope);
+    mEnvelopes[slotIndex].mReadyEnvelopes.push_back(envelope);
+
+    mReadyEnvelopesSize.inc();
+    mSeenEnvelopesSize.inc();
+    return true;
+}
+
+bool
+ReadyEnvelopes::pop(uint64_t slotIndex, SCPEnvelope& ret)
+{
+    auto it = mEnvelopes.begin();
+    while (it != mEnvelopes.end() && it->first <= slotIndex)
+    {
+        auto& v = it->second.mReadyEnvelopes;
+        if (v.size() != 0)
+        {
+            ret = v.back();
+            v.pop_back();
+
+            mReadyEnvelopesSize.dec();
+            traceEnvelope(mApp, "Popped", ret);
+            return true;
+        }
+        it++;
+    }
+    return false;
+}
+
+std::vector<uint64_t>
+ReadyEnvelopes::readySlots()
+{
+    std::vector<uint64_t> result;
+    for (auto const& entry : mEnvelopes)
+    {
+        if (!entry.second.mReadyEnvelopes.empty())
+        {
+            result.push_back(entry.first);
+        }
+    }
+    return result;
+}
+
+void
+ReadyEnvelopes::setMinimumSlotIndex(uint64_t slotIndex)
+{
+    mMinimumSlotIndex = slotIndex;
```

Comment:
not a big fan of `mMinimumSlotIndex` here either: the filtering should happen at herder (that knows which slot it's tracking). It's actually quite error prone to have to keep those mininum/maximum slots in sync.
For example, I think on startup you are not setting `mMinimumSlotIndex` so it stays as 0 even though that's not correct.